### PR TITLE
apps/rehash.c: avoid printf format warning [-Wformat]

### PR DESCRIPTION
**Pull request by mspncp**
The `aarch64-linux-android33-clang` cross-compiler (v14.0.6) complains twice about an unsupported '%n' format specifier, preventing a successful `--strict-warnings` build:

    error: '%n' specifier not supported on this platform [-Werror,-Wformat]
                BIO_snprintf(buf, buflen, "%s%s%n%08x.%s%d",

This is a false positive, because BIO_snprintf() implements its own format parsing (which is implemented in the _dopr() function).

This commit fixes the problem by rewriting the code to dispense with the dubious '%n' format specifier. As a side-effect, the code becomes a little bit more comprehensible and self-explaining.

- _Alternative to pull request #22503_
- _Classified as a bug and marked for backporting, because it prevents a successful --strict-warnings build_
- _Not tested yet_

### CODE REVIEW

File: apps/rehash.c
Comment (paulidale) [inline apps/rehash.c:370]: nit: `dirlen != 0`

Comment (mspncp) [reply apps/rehash.c:370]: I just moved existing code, but sure, I'll change it.

Comment (mspncp) [reply apps/rehash.c]: done.

**Comment by paulidale** [comment]
Noting that the ANSI C standard section 7.19.6.1 paragraph 8 defines:

> n
> The argument shall be a pointer to signed integer into which is written the
> number of characters written to the output stream so far by this call to
> fprintf. No argument is converted, but one is consumed. If the conversion
> specification includes any flags, a field width, or a precision, the behavior is
> undefined.

What is being done here is compliant to the version of C we claim to support.  It's also valid in all later & earlier versions.
IMO: this is a bug in the Android compiler/library **not** in OpenSSL.

**Comment by mspncp** [comment]
> IMO: this is a bug in the Android compiler/library **not** in OpenSSL.

The '%n' specifier is a bit controversial, it's considered unsafe. It seems that Android already removed it more than 10 years ago. And in MSVC it needs to be enabled explicitly.

grafik

https://stackoverflow.com/questions/3401156/what-is-the-use-of-the-n-format-specifier-in-c

**Comment by paulidale** [comment]
Mollycoddling.  There is nothing dangerous about it.  Especially with format / arg checking to ensure you pass an int * in that position.

**Comment by mspncp** [comment]
> Mollycoddling. There is nothing dangerous about it. Especially with format / arg checking to ensure you pass an int * in that position.

I completely agree. But the discussion is pointless: fact is that the option does not exist on Android. The reason why I marked this as a bug is not because I consider '%n' to be unsafe. It's because the compiler warning prevents a successful build with `--strict-warnings`, which I consider a defect.

**Comment by slontis** [comment]
Why is the %n in v3_utl.c not causing the same issue?

**Comment by mspncp** [comment]
> Why is the %n in v3_utl.c not causing the same issue?

You mean this one here?

https://github.com/openssl/openssl/blob/6a0ae393dd554eb718e5148696e8f437d4faae5b/crypto/x509/v3_utl.c#L1159-L1160

Here the '%n' is used with `*scanf()`, not `*printf()`. 

In the `*scanf()` family, most format specifiers have pointer arguments wich receive values. That's probably the reason why it's allowed there and only forbidden in `*printf()`.

**Review by t8m — APPROVED** [review]
IMO this is a good workaround which does not really introduce any overhead.

**Comment by mspncp** [comment]
@openssl/otc how are we going to handle pull requests on master which miss the 3.2 beta release? Wouldn't it be a good idea to add a branch label for 3.2 now?

**Comment by mspncp** [comment]
ping @openssl/committers for second review.

**Review by DDvO — APPROVED** [review]
LGTM

**Comment by DDvO** [comment]
> In the `*scanf()` family, most format specifiers have pointer arguments wich receive values. That's probably the reason why it's still allowed there and was only removed from `*printf()`.

I had not been aware of (or forgotten about) the `%n` format specifier but consider it a useful feature.
A pity that some compilers disallow it, for strange security concerns.

As we know, using a pointer for indirectly receiving a function result is a very common thing.
I guess the fear is that, using `*printf` witih `%n`, one could accidentally get memory overwritten
since the pointer is used in the opposite (output) direction, compared to the most common pointer uses, e.g., with `%s`.

Yet users of `*printf` should know what they are doing.
And those afraid of pointers should not use C.

**Comment by mspncp** [comment]
My guess: it's because of the widespread bad habit of omitting the format string and carelessly calling `printf(untrusted_input)` instead of `printf("%s", untrusted_input)` (which nobody would do with `scanf`). 

So a useful feature turns into a security risk because of bad habits 🤷‍♂️

**Comment by t8m** [comment]
Merged to the master, 3.2, 3.1 and 3.0 branches. Thank you.

**Pull request closed without merging**
